generator.ts

```diff
@@ -0,0 +1,254 @@
+ * @returns The value of operationId
+ */
+const getOperationId = (endpoint: string, resourceName: string, resource: any): string => {
+  if (endpoint === 'get') {
+    return `${endpoint}${_.capitalize(resource.plural)}`;
+  }
+  const words = _.words(endpoint, /(^[a-z]+)|([A-Z][a-z]*)/g);
+  words.splice(1, 0, `${_.capitalize(resourceName)}`);
+  return _.join(words, '');
+};
+
+/**
+ * Creates all resource schemas in the openapi object
+ *
+ * @param config - The config file object
+ * @param openapi - The openapi object
+ * @returns The updated openapi object
+ */
+const buildResources = (config: any, openapi: any) => {
+  _.forEach(config.resources, (resource, resourceName) => {
+    const resourceSchemaPrefix: string = getResourceSchemaPrefix(resourceName);
+
+    const resourceSchema = getResourceSchema(resource, resourceName);
+    const resourceSchemaName = `${resourceSchemaPrefix}Resource`;
+    openapi.components.schemas[resourceSchemaName] = resourceSchema;
+
+    const resultSchema = getResultSchema(resourceSchemaName);
+    openapi.components.schemas[`${resourceSchemaPrefix}Result`] = resultSchema;
+
+    const setResultSchema = getSetResultSchema(resource, resourceSchemaName);
+    openapi.components.schemas[`${resourceSchemaPrefix}SetResult`] = setResultSchema;
+
+    if (_.includes(resource.endpoints, 'post')) {
+      const postBodySchema = getPostBodySchema(resource, resourceSchemaName);
+      openapi.components.schemas[`${resourceSchemaPrefix}PostBody`] = postBodySchema;
+    }
+    if (_.includes(resource.endpoints, 'patchById')) {
+      const patchBodySchema = getPatchBodySchema(resource, resourceSchemaName);
+      openapi.components.schemas[`${resourceSchemaPrefix}PatchBody`] = patchBodySchema;
+    }
+  });
+  return openapi;
+};
+
+/**
+ * Gets the path for an endpoint
+ *
+ * @param endpoint
+ * @param plural - The value of resource.plural from the config file
+ * @returns The path for the endpoint
+ */
+const getPath = (endpoint: string, plural: string): string => {
+  if (_.includes(['get', 'post'], endpoint)) {
+    return `/${plural}`;
+  }
+  if (_.includes(['getById', 'patchById', 'deleteById'], endpoint)) {
+    return `/${plural}/{id}`;
+  }
+  throw Error(`Unexpected endpoint ${endpoint}`);
+};
+
+/**
+ * Gets responses for an endpoint
+ *
+ * @param endpoint
+ * @param resourceSchemaPrefix - The prefix of the resource schema in the openapi document
+ * @returns The responses object
+ */
+const getResponses = (endpoint: string, resourceSchemaPrefix: string): any => {
+  const genericResponse = (code: string) => ({
+    $ref: `#/components/responses/${code}`,
+  });
+
+  const responses: any = {};
+  if (endpoint === 'deleteById') {
+    responses['204'] = {
+      description: 'REPLACEME',
+    };
+  } else {
+    const schemaSuffix = _.includes(['post', 'patchById', 'getById'], endpoint)
+      ? 'Result'
+      : 'SetResult';
+    const code = endpoint === 'post' ? '201' : '200';
```

Comment:
Hmm, I think sometimes a successful `post`/`patch` can also be [`202 - Accepted`](https://jsonapi.org/format/#crud-creating-responses). The most common case is that when we need to call someone else's store procedures and they also need to do some validation to check the data packet we send before inserting/updating data into the DB.

I'm fine to use `201` as the default value for `post`, but I guess we'll probably wanna allow users to customize successful response code from the config file. What's your thought on it?

Reply:
Ah interesting, nice catch. I think a question at the core of several of these issues is if we want the generator script to

1. Only implement logic that comes from a template or function
2. Implement all of the logic in 1 plus all of the logic from the openapi document
3. Something in between

I was originally thinking that the script should have the functionality of 1, meaning anything that is implementation-specific and could be changed in a single place in the openapi file should not be handled by the generator script. This would mean things like `title` and `description` should just have something like `"REPLACEME"` or `""`. The reasoning behind this is that the user would be able to modify this in the openapi file directly rather than in the generator config file.

Under this assumption, I would say that the generator script should always output '201'. Then, if the user needs to change this, they could modify the openapi file directly.

The main reasoning behind not doing 2 or 3 is that we would essentially be creating a subset of the openapi spec. This would be a lot of work and woulldn't provide much benefit, other than making all the openapi changes just in one file. What are your thoughts on this?